test_app/tests/authentication/test_migrate_user_to_authenticator.py

```diff
@@ -0,0 +1,370 @@
+from ansible_base.authentication.models import AuthenticatorUser
+from ansible_base.authentication.utils.authentication import determine_username_from_uid_social, get_or_create_authenticator_user
+from ansible_base.lib.utils.response import get_relative_url
+from test_app.models import ThingSomeoneOwns, ThingSomeoneShares
+
+User = get_user_model()
+
+
+def authenticator_user_factory(uid, username, authenticator, things=[]):
+    user, _ = User.objects.get_or_create(username=username)
+    auth_user = AuthenticatorUser.objects.create(uid=uid, user=user, provider=authenticator)
+
+    for thing in things:
+        ThingSomeoneOwns.objects.create(owner=user, thing=thing)
+
+    return auth_user
+
+
+def get_social_pipeline_kwargs(uid, username, authenticator, **kwargs):
+    return {
+        "uid": uid,
+        "details": {"username": username},
+        "backend": get_authenticator_class(authenticator.type)(database_instance=authenticator),
+        **kwargs,
+    }
+
+
+@pytest.mark.django_db
+def test_migrate_single_user(oidc_authenticator, keycloak_authenticator):
+    oidc_authenticator.auto_migrate_users_to = keycloak_authenticator
+    oidc_authenticator.save()
+
+    oidc_user = authenticator_user_factory("a9a3b88a-7c1a-4fd8-b735-f20c338debcc", "oidc_user", oidc_authenticator).user
+
+    kwargs = get_social_pipeline_kwargs(
+        uid="oidc_user", username="oidc_user", authenticator=keycloak_authenticator, response={"sub": "a9a3b88a-7c1a-4fd8-b735-f20c338debcc"}
+    )
+    username = determine_username_from_uid_social(**kwargs)
+
+    oidc_user.refresh_from_db()
+
+    assert username["username"] == "oidc_user"
+    assert not AuthenticatorUser.objects.filter(provider=oidc_authenticator, user=oidc_user).exists()
+
+
+@pytest.mark.django_db
+def test_migrate_single_user_existing(oidc_authenticator, keycloak_authenticator):
+    oidc_authenticator.auto_migrate_users_to = keycloak_authenticator
+    oidc_authenticator.save()
+
+    shared_thing = ThingSomeoneShares.objects.create(thing="shared")
+    share_user = User.objects.create(username="sharing_is_caring")
+    share_user.things_i_share.add(shared_thing)
+
+    oidc_user = authenticator_user_factory(
+        "a9a3b88a-7c1a-4fd8-b735-f20c338debcc",
+        "oidc_user",
+        oidc_authenticator,
+        things=[
+            "thing1",
+            "conflict_thing",
+        ],
+    ).user
+
+    oidc_user.things_i_share.add(shared_thing)
+
+    kc_user = authenticator_user_factory(
+        "kc_user",
+        "kc_user",
+        keycloak_authenticator,
+        things=[
+            "thing2",
+            "conflict_thing",
+        ],
+    ).user
+
+    kwargs = get_social_pipeline_kwargs(
+        uid="kc_user", username="kc_user", authenticator=keycloak_authenticator, response={"sub": "a9a3b88a-7c1a-4fd8-b735-f20c338debcc"}
+    )
+    username = determine_username_from_uid_social(**kwargs)
```

Comment:
This seems to call `migrate_from_existing_authenticator`, but it's really non-obvious from _this_ name that this is where the new login happens.